library/common/main_interface.h

```diff
@@ -1,12 +1,202 @@
+/**
+ * Called when the async HTTP stream has an error.
+ * @return envoy_error, the error received/caused by the async HTTP stream.
+ */
+typedef envoy_error (*on_error)();
+
+#ifdef __cplusplus
+} // function pointers
+#endif
+
+/**
+ * Interface that can handle HTTP callbacks.
+ */
+typedef struct {
+  on_headers h;
+  on_data d;
+  on_trailers t;
+  on_error e;
+} envoy_observer;
+
+#ifdef __cplusplus
+extern "C" { // functions
+#endif
+
+/**
+ * Open an underlying HTTP stream.
+ * @param observer, the observer that will run the stream callbacks.
+ * @return envoy_stream, with a stream handle and a success status, or a failure status.
+ */
+envoy_stream start_stream(envoy_observer observer);
+
+/**
+ * Send headers over an open HTTP stream. This method can be invoked once and needs to be called
+ * before send_data.
+ * @param stream, the stream to send headers over.
+ * @param headers, the headers to send.
+ * @param end_stream, supplies whether this is headers only.
+ * @return envoy_status_t, the resulting status of the operation.
+ */
+envoy_status_t send_headers(envoy_stream_t stream, envoy_headers headers, bool end_stream);
+
+/**
+ * Send data over an open HTTP stream. This method can be invoked multiple times.
+ * @param stream, the stream to send data over.
+ * @param data, the data to send.
+ * @param end_stream, supplies whether this is the last data in the stream.
+ * @return envoy_status_t, the resulting status of the operation.
+ */
+envoy_status_t send_data(envoy_stream_t stream, envoy_data data, bool end_stream);
+
+/**
+ * Send metadata over an HTTP stream. This method can be invoked multiple times.
+ * @param stream, the stream to send metadata over.
+ * @param metadata, the metadata to send.
+ * @param end_stream, supplies whether this is the last data in the stream.
+ * @return envoy_status_t, the resulting status of the operation.
+ */
+envoy_status_t send_metadata(envoy_stream_t stream, envoy_headers metadata, bool end_stream);
+
+/**
+ * Send trailers over an open HTTP stream. This method can only be invoked once per stream.
+ * Note that this method implicitly ends the stream.
+ * @param stream, the stream to send trailers over.
+ * @param trailers, the trailers to send.
+ * @return envoy_status_t, the resulting status of the operation.
+ */
+envoy_status_t send_trailers(envoy_stream_t stream, envoy_headers trailers);
+
+/**
+ * Half-close an HTTP stream. The stream will be observable and may return further data
+ * via the observer callbacks. However, nothing further may be sent.
+ * @param stream, the stream to close.
+ * @return envoy_status_t, the resulting status of the operation.
+ */
+envoy_status_t close_stream(envoy_stream_t stream);
+
+/**
+ * Detach all observers from a stream and send an interrupt upstream if supported by transport.
+ * @param stream, the stream to evict.
+ * @return envoy_status_t, the resulting status of the operation.
+ */
+envoy_status_t evict_stream(envoy_stream_t stream);
```

Comment:
@goaway nit: should we name start/close/evict_stream, just start/close/evict given that the return value for the first, and the args for the latter two make it evident what is getting started, closed, and evicted?

Reply:
Since the argument is technically still a primitive and there's no overloading by signature, I think it's maybe still better to be explicit in the function name.

That said, I dislike this name now.

What do you think about `locally_close_stream` and `interrupt_stream`?

Reply:
+1 I think those are good.